### PR TITLE
metrics: add footprint density/scaling test

### PR DESCRIPTION
**Pull request by grahamwhaley**
Add the footprint memory/density scaling test.
Add the relevant documentation as well.
Update the json library to enable us to build the results fragments nicely.

### CODE REVIEW

**Comment by egernst** [comment]
@klynnrif @intelkevinputnam - PTAL.

File: metrics/README.md [hunk 1]
```diff
@@ -134,6 +134,7 @@ The preferred API to store results is through the provided JSON API.
 The API provides two groups of functions:
```
Comment (intelkevinputnam) [inline metrics/README.md]: No longer two groups. Am I right?

File: metrics/density/footprint_data.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
+# Footprint data script details
+
+The `footprint_data.sh` script sequentially runs a number of identical containers
+via the docker engine and takes a number of memory related measurements after each
+launch. The script is generally not used in a CI type environment, but is intended
+more for hand run/analysis.
```
Comment (intelkevinputnam) [inline metrics/density/footprint_data.md]: Line 6 - change to "to be run and analyzed manually."

File: metrics/density/footprint_data.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+The script can be configured via a number of environment variables.
```
Comment (intelkevinputnam) [inline metrics/density/footprint_data.md]: Line 8 - change to "can be configured by setting a number of ..."

File: metrics/density/footprint_data.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+Details of the configurable variables are listed below, along with a small example
+invocation script.
+
+## Variables
+The script action can be modified by setting a number of environment variables.
```
Comment (intelkevinputnam) [inline metrics/density/footprint_data.md]: Replace line 14 and the beginning of line 16 with: "Environment variables can take effect in two ways." Add a blank line after that then resume with "Some variables ..."

File: metrics/density/footprint_data.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+| `DUMP_CACHES` | A flag to note if the system caches should be dumped before capturing stats
+| `DATAFILE` | Can be set to over-ride the default JSON results filename
+
+## Output files
+The names of the JSON files generated by the test are dictated by some of the parameters
+the test is utilising. The default filename is generated in the form of:
+`footprint-${PAYLOAD}[-ksm].json`
+
+## Measurements
+The test measures, calculates and stores a number of data items:
```
Comment (intelkevinputnam) [inline metrics/density/footprint_data.md]: Line 44 - add a comma - "... measures, calculates, and stores ..."

File: metrics/density/footprint_data.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+| `cached` | 'Cached' memory from `/proc/meminfo`
+| `smem_free` | Free memory as reported by `smem`
+| `free_decr` | Decrease in Free memory reported by `smem` since start of test
+| `anon` | AnonPages as reported from `/proc/meminfo`
+| `mapped` | Mapped pages as reported from `/proc/meminfo`
+| `cached` | Cached pages as reported from `/proc/meminfo`
+| `slab` | Slab as reported from `/proc/meminfo`
+
+## Example script
+The below script is an example of how to configure the environment variables and
```
Comment (intelkevinputnam) [inline metrics/density/footprint_data.md]: Line 64 - change to: "The following script ..."

File: metrics/README.md (hunk 1 above)
Comment (grahamwhaley) [reply metrics/README.md]: Indeed, you are correct. Reworded to try to make count independent :-)

**Comment by grahamwhaley** [comment]
@intelkevinputnam thx. updates applied. pls re-review.

File: metrics/README.md [hunk 7]
```diff
@@ -131,9 +131,10 @@ we provide a bash API to aid storing results in a uniform manner.
 
 The preferred API to store results is through the provided JSON API.
 
-The API provides two groups of functions:
+The API provides the following groups of functions:
 - A set of functions to init/save the data and add 'top level' JSON fragments
 - A set of functions to construct arrays of JSON fragments, which are then added as a top level fragment when complete
+- A set of functions to construct elements of an array from sub-fragments, and then finalize that element when all fragments have been added.
```
Comment (klynnrif) [inline metrics/README.md]: Suggested rewrite to stay active: 
A set of functions to construct elements of an array from sub-fragments, and then finalize that element when all fragments are added.

File: metrics/README.md [hunk 8]
```diff
@@ -192,6 +193,20 @@ Add a fully formed JSON fragment to the JSON array store.
...
+
+Add a fully formed JSON fragment to the current array element.
+
+| Arg  | Description |
+| ---- | ----------- |
+| json | A fully formed JSON fragment |
+
+#### metrics_json_close_array_element()
+
+Finalize (close) the current array element. This will encorporate
```
Comment (klynnrif) [inline metrics/README.md]: Lines 206-208 suggested rewrite: 
... This incorporates any array_fragment parts into the current array element, closes that array element, and resets the in-flight array_fragment store.

File: metrics/density/footprint_data.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
+# Footprint data script details
+
+The `footprint_data.sh` script runs a number of identical containers sequentially
+via the docker engine and takes a number of memory related measurements after each
```
Comment (klynnrif) [inline metrics/density/footprint_data.md]: through the docker engine ...

File: metrics/density/footprint_data.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+launch. The script is generally not used in a CI type environment, but is intended
+to be run and analyzed manually.
+
+The script can be configured by setting a number of environment variables.
```
Comment (klynnrif) [inline metrics/density/footprint_data.md]: You can configure the script by setting a number of environment variables.

File: metrics/density/footprint_data.md [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+Details of the configurable variables are listed below, along with a small example
```
Comment (klynnrif) [inline metrics/density/footprint_data.md]: The following sections list details of the configurable variables, along with a small example

File: metrics/density/footprint_data.md [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+invocation script.
+
+## Variables
+Environment variables can take effect in two ways.
+
+Some variables affect how the payload is executed. The `RUNTIME` and `PAYLOAD`
+arguments directly affect the payload execution via this line in the script:
```
Comment (klynnrif) [inline metrics/density/footprint_data.md]: arguments directly affect the payload execution with the following line in the script:

**Comment by jodh-intel** [comment]
Once @klynnrif's comments resolved...

lgtm

**Comment by grahamwhaley** [comment]
@klynnrif updates applied - thx

**Pull request merged**